std/experimental/testing/gen_ut_main.d

```diff
@@ -0,0 +1,153 @@
+        "version", "Show version.", &options.showVersion,
+        );
+
+    if (getOptRes.helpWanted)
+    {
+        defaultGetoptPrinter("Usage: gen_ut_main [options] [test1] [test2]...",
+                             getOptRes.options);
+        options.help = true;
+        return options;
+    }
+
+    if (options.showVersion)
+    {
+        writeln("gen_ut_main version v0.2.5");
+        return options;
+    }
+
+    if (options.fileName)
+    {
+        import std.file: exists, remove;
+        if (exists(options.fileName)) remove(options.fileName);
+    }
+    else
+    {
+        options.fileName = createFileName(); //random filename
+    }
+
+    options.dirs = args.length <= 1 ? ["tests"] : args[1..$];
+
+    if (options.verbose)
+    {
+        writeln(__FILE__, ": finding all test cases in ", options.dirs);
+    }
+
+    return options;
+}
+
+private string createFileName()
+{
+    import std.random;
+    import std.ascii : letters, digits;
+    import std.path: buildPath, tempDir;
+
+    immutable nameLength = uniform(10, 20);
+    immutable alphanums = letters ~ digits;
+    auto fileName = "" ~ letters[uniform(0, letters.length)];
+
+    foreach(i; 0 .. nameLength)
+    {
+        fileName ~= alphanums[uniform(0, alphanums.length)];
+    }
+
+    return buildPath(tempDir(),  fileName ~ ".d");
+}
+
+auto findModuleEntries(in string[] dirs)
+{
+    import std.exception: enforce;
+    import std.file: DirEntry, dirEntries, isDir, SpanMode;
+    import std.path: buildNormalizedPath;
+
+    DirEntry[] modules;
+    foreach(dir; dirs)
+    {
+        enforce(isDir(dir), dir ~ " is not a directory name");
+        auto entries = dirEntries(dir, "*.d", SpanMode.depth);
+        auto normalised = entries.map!(a => DirEntry(buildNormalizedPath(a.name)));
+        modules ~= normalised.array;
+    }
+
+    return modules;
+}
+
+auto findModuleNames(in string[] dirs)
+{
+    import std.path: dirSeparator;
+    //cut off extension
+    return findModuleEntries(dirs).map!(a => replace(a.name[0 .. $-2],
+                                                     dirSeparator, ".")).array;
+}
+
+private auto writeFile(in Options options, in string[] modules)
+{
+    writeln("Writing to unit test main file ", options.fileName);
+    writeln("Do not forget to use -unittest when executing ", options.fileName);
+
+    auto wfile = File(options.fileName, "w");
+    wfile.writeln("//Automatically generated by ",
+                  "std.experimental.testing.gen_ut_main, do not edit by hand");
+    wfile.writeln("import std.stdio;");
+    wfile.writeln("import std.experimental.testing;");
+
+    wfile.writeln("");
+    wfile.writeln("int main(string[] args)");
+    wfile.writeln("{");
+    wfile.writeln(`    writeln("\nAutomatically generated file ` ~
+                  options.fileName.replace("\\", "\\\\") ~ `");`);
+    wfile.writeln("    writeln(`Running unit tests from dirs " ~
+                  options.dirs.to!string ~ "`);");
+
```

Comment:
the part starting after File(options.....) untiil this comment could be one q{ } string